심화/Chapter 1/소민/README.md

```diff
@@ -0,0 +1,172 @@
+# DB 부하 줄이기
+
+## 데이터 자체를 쪼개기
+
+### Sharding
+<details><summary>샤딩 구조</summary>
+
+[![샤딩 이미지](./images/sharding.png)](./images/sharding.png)
+
+</details>
+
+하나의 DB를 개별 서버의 샤드로 나누는 것을 의미한다.
+* 샤드란? 데이터를 나눈 작은 단위
+* 즉 데이터를 나눈 작은 단위(샤드)를 개별 DB 서버에서 사용하는 것
+
+같은 스키마(테이블 구조)를 가진 샤드를 여러대로 분리한다는 점이 특징
+* 동일한 테이블을 여러 DB 서버(샤드)로 나누어 트래픽을 분산시키는 수평적 확장
+* 개별 샤드에 동일한 스키마의 중복되지 않는 데이터가 저장되어 있다.
+
+#### 장점
+* 여러 서버에 데이터가 분산 저장되어 쿼리 성능 향상 (필요한 데이터가 있는 샤드에만 접근하므로, 스캔 범위도 줄어들고 병렬적으로 여러 샤드 스캔이 가능함)
+* 개별 샤드에 대해 병렬적으로 RW 가능 -> 쓰기 성능 향상
+* 단일 DB의 용량 문제도 해결할 수 있음 (대용량 데이터 분리) 
+
+#### 단점
+* 전체 데이터에 대한 쿼리 복잡도 향상
+  * e.g., 전체 데이터 개수를 조회하는 쿼리
+* 샤드 간 트랜잭션 관리 복잡도 향상
+
+#### 장애 복구
+* 문제가 생긴 샤드만 복구
+* replication/백업된게 있으면 바로 복구 가능, 그 외에는 다운타임 발생(?)
+
+
+### Partitioning
+
+<details><summary>파티셔닝 구조</summary>
+
+[![파티셔닝 이미지](./images/partitioning.png)](./images/partitioning.png)
+
+</details>
+
+샤딩과 비슷하지만 **하나의 DB 서버 내에서 한 테이블의 데이터를 분리하는 것**을 의미한다.
+* 한 테이블에 있는 다량의 데이터를 여러 파티션으로 나눠서 저장하는 것
+* 개별 파티션에 동일한 스키마, 중복되지 않는 데이터가 존재한다.
+
+#### 장점
+* 필요한 데이터가 있는 파티션만 스캔하여 쿼리 -> 성능 향상
+* 여러 파티션에 대해 병렬적 RW 작업 가능 -> 쓰기 성능 향상
+* 단일 DB 서버 한 대로 데이터 분산 가능
+
+#### 단점
+* 샤딩과 마찬가지로, 파티션을 나눌 키/전략 설정이 필요함
+* 하나의 파티션에 데이터가 몰리는 불균형 발생 가능
+* 여러 파티션에 걸쳐있는 데이터를 조회해야 하는 쿼리 복잡도 향상
+
+#### 장애 복구
+* 백업이 있으면 해당 백업본으로 복구 -> 데이터 양이 많다면 복구가 오래걸릴 수 있음
+* MySQL 등 일부 DBMS에서는 파티션에 대한 replication도 제공해서 바로 복구 가능
```

Comment:
파티션에 대한 replication 은 어떻게 저장되고 있는거지...😵

Reply:
MySQL에서 지원하는 파티션 Replication이 정확히 어떻게 지원되는지에(어떤 형태로 구성되는지) 대한 문서는 못찾았는데.. 공식문서에 있는 내용을 보면 아래 구조인 것으로 생각됨.

```
master server - N 파티션
slave server - N 파티션(master와 동일한 데이터 분배/파티션 구조로 구성되어야만 함)
```

> Server SQL modes also impact replication of partitioned tables. Disparate SQL modes on source and replica can lead to partitioning expressions being evaluated differently; this can cause the distribution of data among partitions to be different in the source's and replica's copies of a given table, and may even cause inserts into partitioned tables that succeed on the source to fail on the replica. For best results, you should always use the same server SQL mode on the source and on the replica.

> 요약: Server Mode SQL은 파티션의 replication에도 영향을 미칠 수 있다. source(master)와 replica(slave) 파티션에 데이터가 균등하게 분배되도록 구성해야한다.
- [MySQL Partitioning limitation](https://dev.mysql.com/doc/refman/8.0/en/partitioning-limitations.html) -> 위 내용에 대한 원본 문서
- [MySQL Replication and Partitioning](https://dev.mysql.com/doc/refman/8.4/en/replication-features-partitioning.html) -> MySQL에서 파티션에 대해 replication이 가능하다는 내용